File: apps/agent/app/agent/tools/send_receipt.py
"""Tool contract and implementation: send_receipt."""

from typing import Any, Literal
import asyncpg
from pydantic import BaseModel, Field
from app.config import settings
from app.db.client import with_clinic, log_audit
from app.llm.base import ToolSpec

TOOL_NAME = "send_receipt"
TOOL_DESCRIPTION = "Issue and dispatch a payment receipt or invoice document link to a patient."

ConversationChannel = Literal["whatsapp", "voice", "instagram", "facebook", "google"]
FileKind = Literal["xray", "photo", "document", "receipt"]


class SendReceiptInput(BaseModel):
    """Input parameters for send_receipt tool."""

    patient_id: str = Field(..., description="UUID of the patient")
    invoice_id: str = Field(..., description="UUID or identifier of the invoice/receipt to dispatch")
    channel: ConversationChannel = Field(
        ...,
        description="Delivery channel (matches DB conversation_channel enum: whatsapp, voice, instagram, facebook, google)",
    )


class SendReceiptOutput(BaseModel):
    """Output results for send_receipt tool."""

    receipt_id: str
    file_kind: FileKind = "receipt"
    sent: bool = True
    delivery_channel: ConversationChannel
    message: str = "Receipt link dispatched to patient."


TOOL_SPEC = ToolSpec(
    name=TOOL_NAME,
    description=TOOL_DESCRIPTION,
    parameters=SendReceiptInput.model_json_schema(),
)


async def execute(input_data: SendReceiptInput, context: dict[str, Any]) -> SendReceiptOutput:
    """Verify invoice and receipt PDF in clinic scope, log outbox message, and record audit log."""
    clinic_id = context.get("clinic_id") or settings.default_clinic_id
    patient_phone = context.get("sender") or "+923001234500"

    async def _execute(conn: asyncpg.Connection) -> SendReceiptOutput:
        # 1. Verify invoice belongs to the patient and clinic
        invoice_row = await conn.fetchrow(
            """
            SELECT id, invoice_number, patient_id
            FROM invoices
            WHERE id = $1::uuid AND clinic_id = $2::uuid AND patient_id = $3::uuid
            LIMIT 1
            """,
            input_data.invoice_id,
            clinic_id,
            input_data.patient_id,
        )
        if not invoice_row:
            raise ValueError(
                f"Invoice '{input_data.invoice_id}' does not exist or does not belong to specified patient/clinic."
            )

        # 2. Verify receipt exists with storage_key
        receipt_row = await conn.fetchrow(
            """
            SELECT id, receipt_number, storage_key
            FROM receipts
            WHERE invoice_id = $1::uuid AND clinic_id = $2::uuid
            LIMIT 1
            """,
            input_data.invoice_id,
            clinic_id,
        )
        if not receipt_row or not receipt_row["storage_key"]:
            raise ValueError(
                f"Receipt with generated PDF storage key not found for invoice '{input_data.invoice_id}'."
            )

        receipt_id = str(receipt_row["id"])
        storage_key = receipt_row["storage_key"]

        # 3. Dispatch receipt link to dev_outbox (mock adapter)
        body = f"Receipt link: {storage_key}"
        await conn.execute(
            """
            INSERT INTO dev_outbox (
                id, channel, "from", "to", body, provider, direction, created_at
            ) VALUES (
                gen_random_uuid(), $1, 'system', $2, $3, 'mock', 'outbound', NOW()
            )
            """,
            input_data.channel,
            patient_phone,
            body,
        )

        # 4. Write immutable audit log
        await log_audit(
            conn=conn,
            clinic_id=clinic_id,
            action="receipt.dispatched",
            entity="receipt",
            entity_id=receipt_id,
            meta={
                "invoice_id": input_data.invoice_id,
                "channel": input_data.channel,
                "storage_key": storage_key,
            },
        )

        return SendReceiptOutput(
            receipt_id=receipt_id,
            file_kind="receipt",
            sent=True,
            delivery_channel=input_data.channel,
            message="Receipt link dispatched to patient.",
        )

    return await with_clinic(clinic_id, _execute)

File: apps/agent/src/agent/tools/send-receipt.ts
import { z } from 'zod';
import { eq, and } from 'drizzle-orm';
import { invoices, receipts, patients, devOutbox } from '@dental-pms/db/schema';
import type { ToolSpec } from '../../llm/base';
import { runInClinic, getClinicIdFromContext } from '../../db/context';
import { writeAuditLog } from '../../db/audit';

export const TOOL_NAME = 'send_receipt';
export const TOOL_DESCRIPTION =
  'Issue and dispatch a payment receipt or invoice document link to a patient.';

export const sendReceiptInputSchema = z.object({
  patient_id: z.string().uuid().describe('UUID of the patient'),
  invoice_id: z.string().uuid().describe('UUID or identifier of the invoice/receipt to dispatch'),
  channel: z
    .enum(['whatsapp', 'voice', 'instagram', 'facebook', 'google'])
    .describe('Delivery channel'),
});

export type SendReceiptInput = z.infer<typeof sendReceiptInputSchema>;

export interface SendReceiptOutput {
  receipt_id: string;
  file_kind: 'xray' | 'photo' | 'document' | 'receipt';
  sent: boolean;
  delivery_channel: 'whatsapp' | 'voice' | 'instagram' | 'facebook' | 'google';
  message: string;
}

export const TOOL_SPEC: ToolSpec = {
  name: TOOL_NAME,
  description: TOOL_DESCRIPTION,
  parameters: {
    type: 'object',
    properties: {
      patient_id: { type: 'string', description: 'UUID of the patient' },
      invoice_id: { type: 'string', description: 'UUID or identifier of the invoice/receipt to dispatch' },
      channel: {
        type: 'string',
        enum: ['whatsapp', 'voice', 'instagram', 'facebook', 'google'],
        description: 'Delivery channel',
      },
    },
    required: ['patient_id', 'invoice_id', 'channel'],
  },
};

export async function execute(
  input: SendReceiptInput,
  context?: Record<string, unknown>
): Promise<SendReceiptOutput> {
  const clinicId = getClinicIdFromContext(context);

  return await runInClinic(clinicId, async (tx) => {
    // 1. Verify invoice belongs to patient and clinic
    const [invoice] = await tx
      .select({ id: invoices.id, patientId: invoices.patientId })
      .from(invoices)
      .where(
        and(
          eq(invoices.id, input.invoice_id),
          eq(invoices.patientId, input.patient_id),
          eq(invoices.clinicId, clinicId)
        )
      )
      .limit(1);

    if (!invoice) {
      throw new Error(
        `Invoice ${input.invoice_id} not found for patient ${input.patient_id} in this clinic`
      );
    }

    // 2. Verify receipt exists with storage key
    const [receipt] = await tx
      .select({ id: receipts.id, storageKey: receipts.storageKey })
      .from(receipts)
      .where(
        and(
          eq(receipts.invoiceId, input.invoice_id),
          eq(receipts.clinicId, clinicId)
        )
      )
      .limit(1);

    if (!receipt || !receipt.storageKey) {
      throw new Error(`No receipt document with storage_key found for invoice ${input.invoice_id}`);
    }

    // 3. Resolve patient destination phone
    const [patient] = await tx
      .select({ phone: patients.phone })
      .from(patients)
      .where(and(eq(patients.id, input.patient_id), eq(patients.clinicId, clinicId)))
      .limit(1);

    const recipient = patient?.phone || input.patient_id;

    // 4. Log outbound dispatch into dev_outbox table (Mock delivery)
    await tx.insert(devOutbox).values({
      channel: input.channel,
      from: 'system',
      to: recipient,
      body: `Receipt link: ${receipt.storageKey}`,
      provider: 'mock',
      direction: 'outbound',
      metadata: JSON.stringify({ invoice_id: input.invoice_id, receipt_id: receipt.id }),
    });

    // 5. Write immutable audit log
    await writeAuditLog(tx, clinicId, {
      action: 'receipt.dispatched',
      entity: 'receipt',
      entityId: receipt.id,
      meta: {
        invoice_id: input.invoice_id,
        channel: input.channel,
        recipient,
      },
    });

    return {
      receipt_id: receipt.id,
      file_kind: 'receipt',
      sent: true,
      delivery_channel: input.channel,
      message: `Payment receipt dispatched via ${input.channel}.`,
    };
  });
}
